base/compiler/ssair/passes.jl

```diff
@@ -1,3 +1,416 @@
+        nleaves == nuses_total || continue
+        # Find the type for this allocation
+        defexpr = ir[SSAValue(idx)]
+        isexpr(defexpr, :new) || continue
+        typ = defexpr.typ
+        if isa(typ, UnionAll)
+            typ = unwrap_unionall(typ)
+        end
+        # Could still end up here if we tried to setfield! and immutable, which would
+        # error at runtime, but is not illegal to have in the IR.
+        typ.mutable || continue
+        # Partition defuses by field
+        fielddefuse = SSADefUse[SSADefUse() for _ = 1:fieldcount(typ)]
+        ok = true
+        for use in defuse.uses
+            field = try_compute_fieldidx(typ, ir[SSAValue(use)])
+            field === nothing && (ok = false; break)
+            push!(fielddefuse[field].uses, use)
+        end
+        ok || continue
+        for use in defuse.defs
+            field = try_compute_fieldidx(typ, ir[SSAValue(use)])
+            field === nothing && (ok = false; break)
+            push!(fielddefuse[field].defs, use)
+        end
+        ok || continue
+        preserve_uses = IdDict{Int, Vector{Any}}((idx=>Any[] for idx in IdSet{Int}(defuse.ccall_preserve_uses)))
+        # Everything accounted for. Go field by field and perform idf
+        for (fidx, du) in pairs(fielddefuse)
+            ftyp = fieldtype(typ, fidx)
+            if !isempty(du.uses)
+                push!(du.defs, idx)
+                ldu = compute_live_ins(ir.cfg, du)
+                phiblocks = []
+                if !isempty(ldu.live_in_bbs)
+                    phiblocks = idf(ir.cfg, ldu, domtree)
+                end
+                phinodes = IdDict{Int, SSAValue}()
+                for b in phiblocks
+                    n = PhiNode()
+                    phinodes[b] = insert_node!(ir, first(ir.cfg.blocks[b].stmts), ftyp, n)
+                end
+                # Now go through all uses and rewrite them
+                allblocks = sort(vcat(phiblocks, ldu.def_bbs))
+                for stmt in du.uses
+                    ir[SSAValue(stmt)] = compute_value_for_use(ir, domtree, allblocks, du, phinodes, fidx, stmt)
+                end
+                if !isbits(fieldtype(typ, fidx))
+                    for (use, list) in preserve_uses
+                        push!(list, compute_value_for_use(ir, domtree, allblocks, du, phinodes, fidx, use))
+                    end
+                end
+                for b in phiblocks
+                    for p in ir.cfg.blocks[b].preds
+                        n = ir[phinodes[b]]
+                        push!(n.edges, p)
+                        push!(n.values, compute_value_for_block(ir, domtree,
+                            allblocks, du, phinodes, fidx, p))
+                    end
+                end
+            end
+            for stmt in du.defs
+                stmt == idx && continue
+                ir[SSAValue(stmt)] = nothing
```

Comment:
`Expr(:new)` has some possible side-effects (type checking) which don't appear to be captured above. Perhaps it would be better to handle unused-alloc-elim in a separate pass? For example, see how this is handled now:
```julia
julia> struct A
         x::Int
         A(x) = new(x)
       end

julia> A(3.0)
A(3)

julia> struct B; end

julia> Base.convert(::Type{Int}, b::B) = b

julia> c(x) = A(x).x

julia> c(B())
ERROR: TypeError: in Type, in new, expected Int64, got B

julia> @code_typed c(B())
CodeInfo(:(begin
      # meta: location REPL[1] Type 3
      Core.SSAValue(5) = ($(QuoteNode(B())) isa Int64)::Bool
      Core.SSAValue(6) = (Core.Intrinsics.not_int)(Core.SSAValue(5))::Bool
      unless Core.SSAValue(6) goto 6
      $(Expr(:foreigncall, :(:jl_type_error_new_expr), Union{}, svec(Any, Any), :(:ccall), 2, Int64, :($(QuoteNode(B())))))
      6: 
      # meta: pop location
      return $(QuoteNode(B()))
  end)) => Int64
```